client/extensions/woocommerce/app/products/index.js

```diff
@@ -55,17 +74,42 @@ class Products extends Component {
 		const newSiteId = newProps.site && newProps.site.ID || null;
 		const oldSiteId = site && site.ID || null;
 		if ( oldSiteId !== newSiteId ) {
+			this.setState( { isSearching: false, query: '' } );
 			this.props.fetchProducts( newSiteId, 1 );
 		}
 	}
 
 	switchPage = ( page ) => {
 		const { site } = this.props;
-		this.props.fetchProducts( site.ID, page );
+		const { isSearching } = this.state;
+
+		if ( isSearching ) {
+			this.props.fetchProductSearchResults( site.ID, page );
+		} else {
+			this.props.fetchProducts( site.ID, page );
+		}
+	}
+
+	onSearch = ( query ) => {
+		const { site } = this.props;
+
+		if ( trim( query ) === '' ) {
+			this.setState( { isSearching: false, query: '' } );
+			this.props.clearProductSearch( site.ID );
+			return;
+		}
+
+		this.setState( { isSearching: true, query } );
```

Comment:
Seems like isSearching is always true at the same time query is non trivial. Maybe you only need the query in state and could Boolean( trim( query ) ) elsewhere.  See also my suggestion however on treating "not searching" as a special case of searching with a "trivial" query.